example_buffer.py

```diff
@@ -0,0 +1,198 @@
+
+    def update(self, new_games, samples_per_game=4):
+        """
+        new_games is list of .tfrecord.zz files of new games
+        """
+        new_games.sort(key=os.path.basename)
+        first_new_game = None
+        for i, game in enumerate(tqdm(new_games)):
+            t = file_timestamp(game)
+            if t <= self.examples[-1][0]:
+                continue
+            elif first_new_game is None:
+                first_new_game = i
+                print("Found {}/{} new games".format(len(new_games)-i, len(new_games)))
+
+            choices = [(t, ex) for ex in pick_examples_from_tfrecord(
+                game, samples_per_game)]
+            self.examples.extend(choices)
+
+    def flush(self, path):
+        with timer("Writing examples to " + path):
+            preprocessing.write_tf_examples(
+                path, [ex[1] for ex in self.examples])
+
+    @property
+    def count(self):
+        return len(self.examples)
+
+    def __str__(self):
+        return "ExampleBuffer: {} positions sampled from {} to {}".format(
+            self.count,
+            _ts_to_str(self.examples[0][0]),
+            _ts_to_str(self.examples[-1][0]))
+
+
+def files_for_model(model):
+    return tf.gfile.Glob(os.path.join(LOCAL_DIR, model[1], '*.zz'))
+
+
+def smart_rsync(from_model_num=0, source_dir=rl_loop.SELFPLAY_DIR, dest_dir=LOCAL_DIR):
+    from_model_num = 0 if from_model_num < 0 else from_model_num
+    models = [m for m in rl_loop.get_models() if m[0] >= from_model_num]
+    for m in models:
+        _rsync_dir(os.path.join(
+            source_dir, m[1]), os.path.join(dest_dir, m[1]))
+
+
+def _rsync_dir(source_dir, dest_dir):
+    _ensure_dir_exists(dest_dir)
+    with open('.rsync_log', 'ab') as rsync_log:
+        subprocess.call(['gsutil', '-m', 'rsync', source_dir, dest_dir],
+                        stderr=rsync_log)
+
+
+def fill_and_wait(bufsize=dual_net.EXAMPLES_PER_GENERATION,
+                  write_dir=rl_loop.GOLDEN_CHUNK_DIR,
+                  model_window=100,
+                  threads=8,
+                  skip_first_rsync=False):
+    buf = ExampleBuffer(bufsize)
+    models = rl_loop.get_models()[-model_window:]
+    # Last model is N.  N+1 is training.  We should gather games for N+2.
+    chunk_to_make = os.path.join(rl_loop.GOLDEN_CHUNK_DIR, str(
+        models[-1][0] + 2) + '.tfrecord.zz')
+    while tf.gfile.Exists(chunk_to_make):
+        print("Next chunk ({}) already exists.  Sleeping.".format(chunk_to_make))
+        time.sleep(5*60)
+        models = rl_loop.get_models()[-model_window:]
+    print("Making chunk:", chunk_to_make)
+    if not skip_first_rsync:
+        with timer("Rsync"):
+            smart_rsync(models[-1][0] - 6)
+    files = list(tqdm(map(files_for_model, models), total=len(models)))
+    buf.parallel_fill(list(itertools.chain(*files)), threads=threads)
+
+    print("Filled buffer, watching for new games")
+    while rl_loop.get_latest_model()[0] == models[-1][0]:
```

Comment:
I see, so this continually rsync's and updates the latest buffer until a new model is promoted; it then dumps its contents for training. Seems reasonable to me. The name of this method could probably be more descriptive, but I'm coming up blank.

Reply:
Correct.  I guess i should comment this, eh?